group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,959 @@
+            return this;
+        }
+
+        public CoordinatorRuntime<S, U> build() {
+            if (logContext == null)
+                logContext = new LogContext();
+            if (eventProcessor == null)
+                throw new IllegalArgumentException("Event processor must be set.");
+            if (partitionWriter == null)
+                throw new IllegalArgumentException("Partition write must be set.");
+            if (loader == null)
+                throw new IllegalArgumentException("Loader must be set.");
+            if (coordinatorBuilderSupplier == null)
+                throw new IllegalArgumentException("State machine supplier must be set.");
+
+            return new CoordinatorRuntime<>(
+                logContext,
+                eventProcessor,
+                partitionWriter,
+                loader,
+                coordinatorBuilderSupplier
+            );
+        }
+    }
+
+    /**
+     * The various state that a coordinator for a partition can be in.
+     */
+    enum CoordinatorState {
+        /**
+         * Initial state when a coordinator is created.
+         */
+        INITIAL,
+
+        /**
+         * The coordinator is being loaded.
+         */
+        LOADING,
+
+        /**
+         * The coordinator is active and can service requests.
+         */
+        ACTIVE,
+
+        /**
+         * The coordinator is closed.
+         */
+        CLOSED,
+
+        /**
+         * The coordinator has failed.
+         */
+        FAILED
+    }
+
+    /**
+     * CoordinatorContext holds all the metadata around a coordinator state machine.
+     */
+    class CoordinatorContext {
+        /**
+         * The topic partition backing the coordinator.
+         */
+        final TopicPartition tp;
+
+        /**
+         * The snapshot registry backing the coordinator.
+         */
+        final SnapshotRegistry snapshotRegistry;
+
+        /**
+         * The deferred event queue used to park events waiting
+         * on records to be committed.
+         */
+        final DeferredEventQueue deferredEventQueue;
+
+        /**
+         * The current state.
+         */
+        volatile CoordinatorState state;
+
+        /**
+         * The actual state machine.
+         */
+        volatile S coordinator;
+
+        /**
+         * The current epoch of the coordinator. This represents
+         * the epoch of the partition leader.
+         */
+        volatile int epoch;
+
+        /**
+         * The last offset written to the partition.
+         */
+        volatile long lastWrittenOffset;
+
+        /**
+         * The last offset committed. This represents the high
+         * watermark of the partition.
+         */
+        volatile long lastCommittedOffset;
+
+        /**
+         * Constructor.
+         *
+         * @param tp The topic partition of the coordinator.
+         */
+        private CoordinatorContext(
+            TopicPartition tp
```

Comment:
does this mean a coordinator will be associated with multiple coordinator contexts since a coordinator owns multiple partitions?

never mind -- looks like we have 1-1 mapping of coordinator to topic partition

Reply:
I was also wondering this, so thanks for bringing it up.

Reply:
Correct. Every partition is mapped to a CondinatorContext and a Coordinator. The Coordinator here is basically the replicated state machine.